src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -910,8 +910,12 @@ void IHttp2StreamLifetimeHandler.OnStreamCompleted(int streamId)
                         Input.CancelPendingRead();
                     }
 
-                    // Complete the task waiting on all streams to finish
-                    _streamsCompleted.TrySetResult(null);
+
+                    if (_state != Http2ConnectionState.Open)
```

Comment:
Nice catch @Tratcher. Let's add a new regression unit test that:

1. Opens a stream where the application doesn't block.
2. Closes the stream.
3. Opens a new stream where the application blocks.
4. Closes the input pipe
5. Verifies that Http2Connection.ProcessRequestsAsync doesn't complete until the application for the second stream completes.